### DISCUSSION TITLE
narrowing type resolves to unexpected type

### DISCUSSION THREAD

**Discussion by rchl** [microsoft/pyright#6301] [Q&A] [answered]
I have a home brewed Promise implementation that I've been using for a while and I've noticed that at some point Pyright started showing a new issue with it that wasn't reported before. The types in this implementation have developed to the point that I don't understand them myself anymore so I have to ask whether the current error is expected or not.

This is a very stripped down code that shows the error:

```py
from typing import Callable, Generic, List, Optional, Protocol, Tuple, TypeVar, Union

T = TypeVar('T')
T_contra = TypeVar('T_contra', contravariant=True)
TResult = TypeVar('TResult')


class ResolveFunc(Protocol[T_contra]):
    def __call__(self, resolve_value: T_contra) -> None:
        ...


FullfillFunc = Callable[[T], Union[TResult, 'Promise[TResult]']]


class Promise(Generic[T]):

    def then(self, onfullfilled: FullfillFunc[T, TResult]) -> 'Promise[TResult]':

        def callback_wrapper(resolve_fn: ResolveFunc[TResult], resolve_value: T) -> None:
            result = onfullfilled(resolve_value)
            if isinstance(result, Promise):
                result.then(lambda value: resolve_fn(value))
            else:
                resolve_fn(result)

                pass

        return Promise()
```

The error is:

```
/usr/local/workspace/py-test/foo.py:23:54 - error:
Argument of type "None" cannot be assigned to parameter "resolve_value" of type "TResult@then" in function "__call__"
    Type "None" cannot be assigned to type "TResult@then" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

This is line 23:
```
:23                result.then(lambda value: resolve_fn(value))
```

**Comment by rchl** [comment]
To add some extra bit of detail...

The type of `result` is this before the `isinstance` narrowing:

![Screenshot 2023-11-02 at 17 18 04](https://github.com/microsoft/pyright/assets/153197/2b6b5c20-1c3f-4a44-b299-3b9ec6b892d6)

but after narrowing it becomes:

![Screenshot 2023-11-02 at 17 19 04](https://github.com/microsoft/pyright/assets/153197/ed4eac82-e4b5-4d89-81ef-d7543c9af9f0)

And it feels like `Promise[Unknown]*` is unexpected but then I don't even know what it means exactly with that `*` symbol.

**Reply by erictraut** [reply, marked as answer]
Part of this behavior is expected. The issue is that `result` is of type `TResult | Promise[TResult]`. When you apply the `isinstance` type guard, either of the types in this union could match type `Promise`. It becomes narrowed to `Promise[Unknown] | Promise[TResult]`. 

I see two issues that I'd like to investigate further:
1. Why is the lambda type inference producing a type of `None` for `value`?
2. It appears that the lambda type inference behavior depends on the order of subtypes in the union. If you reverse the order in the definition for `FullfillFunc`, the error goes away. Union ordering shouldn't change the results, so this should be fixed.

I'll create an issue from this discussion thread.
